### PR TITLE
Continuing cleanup.

### PR DESCRIPTION
**Pull request by znado**
- refactoring dataset iterators to return Dict[str, Any] instead of a tuple/triplet of inputs/labels/masks
- cleaning up some of the WMT code
- standardizing param shape/type utilities
- small cleanups/fixes to get all reference submissions to work with random inputs for 1 train and 2 eval steps (takes ~272s total to test)

### CODE REVIEW

File: tests/reference_submission_tests.py [hunk 1]
```diff
@@ -30,51 +33,132 @@
...
-    }
+_EXPECTED_METRIC_NAMES = {
+    'mnist': ['validation/accuracy', 'test/accuracy'],
+    'imagenet': ['validation/accuracy'],
+    'librispeech': ['validation/word_error_rate', 'train/word_error_rate'],
+    'ogbg': [
+        'train/accuracy', 'validation/loss', 'test/mean_average_precision'
+    ],
+    'wmt': ['train/bleu', 'validation/loss', 'test/accuracy'],
 }
```
Comment (fsschneider) [inline tests/reference_submission_tests.py:44]: I might be missing something here, but shouldn't we get more than this?
For `mnist`, for example, do we not get the `loss` and `accuracy` for `train`, `valid`, and `test`?

**Review by fsschneider — COMMENTED** [review]
Looks really awesome!

I added a few comments and nits. I think, after merging #66 and addressing these comments, this is ready to merge.

File: tests/reference_submission_tests.py [hunk 2]
```diff
@@ -102,6 +187,12 @@ def _test_submission(workload_name,
   data_selection = submission_module.data_selection
   get_batch_size = submission_module.get_batch_size
   global_batch_size = get_batch_size(workload_name)
+  global_batch_size = 2
```
[the remark this answers is not in the document]
Comment (znado) [reply tests/reference_submission_tests.py:194]: I want to keep this in there just to make sure that we exercise `get_batch_size`

File: algorithmic_efficiency/workloads/wmt/wmt_jax/models.py [hunk 3]
```diff
@@ -12,17 +12,20 @@
...
+  # num_heads: int = 16
+  # num_layers: int = 6
+  # qkv_dim: int = 1024
+  # mlp_dim: int = 4096
   max_len: int = 256
+  emb_dim: int = 102
+  num_heads: int = 6
+  num_layers: int = 2
+  qkv_dim: int = 102
+  mlp_dim: int = 409
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/workloads/wmt/wmt_jax/models.py]: yea this was just for debugging, didn't mean to commit it. undone!

File: algorithmic_efficiency/workloads/imagenet/imagenet_pytorch/workload.py [hunk 4]
```diff
@@ -126,14 +102,16 @@ def _build_dataset(self,
         pin_memory=True,
         drop_last=is_train)
 
-    if is_train:
-      dataloader = cycle(dataloader)
+    dataloader = cycle(dataloader)
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/workloads/imagenet/imagenet_pytorch/workload.py:117]: I added it so that we can iterate through them across multiple evals. I'm not sure if this is the best way to do it.

File: RULES.md [hunk 5]
```diff
@@ -146,9 +146,8 @@ update_params(
     current_param_container: ParameterContainer,
     current_params_types: ParameterTypeTree,
     model_state: ModelAuxiliaryState,
-    hyperparameters: Hyperparamters,
-    input_batch: Dict[Tensor],
-    label_batch: Dict[Tensor],
+    hyperparameters: Hyperparameters,
+    batch: Dict[str, Tensor],
```
[the remark this answers is not in the document]
Comment (znado) [reply RULES.md:150]: Done.

File: .github/workflows/linting.yml [hunk 6]
```diff
@@ -19,6 +19,7 @@ jobs:
       run: |
         pylint algorithmic_efficiency
         pylint baselines
+        pylint reference_submissions
```
[the remark this answers is not in the document]
Comment (znado) [reply .github/workflows/linting.yml:22]: Yeah I'm open to renaming them, but lets do that in a follow-up PR

File: tests/reference_submission_tests.py (hunk 1 above)
Comment (znado) [reply tests/reference_submission_tests.py]: yep, I added a few more to better cover the cross product of {metric names, split names}

File: algorithmic_efficiency/param_utils.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        param_types_dict[name] = spec.ParameterType.BIAS
+      elif 'BatchNorm' in name:
+        param_types_dict[name] = spec.ParameterType.BATCH_NORM
+      elif 'Conv' in name:
+        param_types_dict[name] = spec.ParameterType.CONV_WEIGHT
+      # Note that this is exact equality, not contained in, because
+      # flax.linen.Embed names the embedding parameter "embedding"
+      # https://github.com/google/flax/blob/main/flax/linen/linear.py#L604.
+      elif name == 'embedding':
+        param_types_dict[name] = spec.ParameterType.EMBEDDING
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/param_utils.py:39]: yes. is the way I added the right way to check the name of an nn.Embedding layer in Pytorch? we should add a test to make sure (for jax and pytorch) these are properly set (at least, for some selection of params). I added a TODO to my list of things to follow up with.

File: algorithmic_efficiency/workloads/librispeech/librispeech_pytorch/workload.py [hunk 8]
```diff
@@ -1,28 +1,31 @@
...
+import torch.utils.data as data_utils
 
+from algorithmic_efficiency import param_utils
 from algorithmic_efficiency import spec
 from algorithmic_efficiency.workloads.librispeech.librispeech_pytorch import \
     input_pipeline
 from algorithmic_efficiency.workloads.librispeech.librispeech_pytorch import \
     models
 
+device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/workloads/librispeech/librispeech_pytorch/workload.py]: done.
